File: golang/07-switch-factory/playlist_curator.go
package kata

import "sort"

type Track struct {
	Title        string
	Tempo        int
	Energy       int
	Instrumental bool
	Danceability int
}

type PlaylistCurator struct{}

func NewPlaylistCurator() *PlaylistCurator {
	return &PlaylistCurator{}
}

func (pc *PlaylistCurator) create_playlist(mood string, tracks []Track) []Track {
	result := []Track{}
	switch mood {
	case "happy":
		for _, t := range tracks {
			if t.Tempo > 120 {
				result = append(result, t)
			}
		}
		sort.Slice(result, func(i, j int) bool { return result[i].Tempo > result[j].Tempo })
		if len(result) > 20 {
			result = result[:20]
		}
	case "sad":
		for _, t := range tracks {
			if t.Tempo < 90 {
				result = append(result, t)
			}
		}
		sort.Slice(result, func(i, j int) bool { return result[i].Tempo < result[j].Tempo })
		if len(result) > 15 {
			result = result[:15]
		}
	case "workout":
		for _, t := range tracks {
			if t.Tempo > 130 && t.Energy > 7 {
				result = append(result, t)
			}
		}
		if len(result) > 25 {
			result = result[:25]
		}
	case "focus":
		for _, t := range tracks {
			if t.Instrumental {
				result = append(result, t)
			}
		}
		if len(result) > 30 {
			result = result[:30]
		}
	case "party":
		for _, t := range tracks {
			if t.Tempo > 110 && t.Danceability > 6 {
				result = append(result, t)
			}
		}
		if len(result) > 20 {
			result = result[:20]
		}
	default:
		result = append(result, tracks...)
		if len(result) > 10 {
			result = result[:10]
		}
	}
	return result
}

File: javascript/07-switch-factory/src/playlist_curator.js
class PlaylistCurator {
  constructor() {}

  create_playlist(mood, tracks) {
    let result = [];

    switch (mood) {
      case 'happy':
        for (let i = 0; i < tracks.length; i++) {
          if (tracks[i].tempo > 120) {
            result.push(tracks[i]);
          }
        }
        result.sort((a, b) => b.tempo - a.tempo);
        if (result.length > 20) {
          result = result.slice(0, 20);
        }
        break;
      case 'sad':
        for (let i = 0; i < tracks.length; i++) {
          if (tracks[i].tempo < 90) {
            result.push(tracks[i]);
          }
        }
        result.sort((a, b) => a.tempo - b.tempo);
        if (result.length > 15) {
          result = result.slice(0, 15);
        }
        break;
      case 'workout':
        for (let i = 0; i < tracks.length; i++) {
          if (tracks[i].tempo > 130 && tracks[i].energy > 7) {
            result.push(tracks[i]);
          }
        }
        if (result.length > 25) {
          result = result.slice(0, 25);
        }
        break;
      case 'focus':
        for (let i = 0; i < tracks.length; i++) {
          if (tracks[i].instrumental === true) {
            result.push(tracks[i]);
          }
        }
        if (result.length > 30) {
          result = result.slice(0, 30);
        }
        break;
      case 'party':
        for (let i = 0; i < tracks.length; i++) {
          if (tracks[i].tempo > 110 && tracks[i].danceability > 6) {
            result.push(tracks[i]);
          }
        }
        if (result.length > 20) {
          result = result.slice(0, 20);
        }
        break;
      default:
        for (let i = 0; i < tracks.length; i++) {
          result.push(tracks[i]);
        }
        if (result.length > 10) {
          result = result.slice(0, 10);
        }
    }

    return result;
  }
}

export { PlaylistCurator };
